---
title: "Exploration_data"
author: "Tina Gorsek Sparovec"
params:
  data: "../data"
  results: "../2.0_results"
  lib: "../0.0_lib"
  maxcores: 6
output:
  html_notebook:
  theme: spacelab
  highlight: textmate
  toc: yes
  number_sections: true
  toc_depth: 3
  toc_float: true
---

Workflow sources: 
[Analyzing RNA-seq data with DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat)

```{r}
library(tidyverse)
library(DESeq2)
library(IHW)
library(pcaExplorer)
library(clusterProfiler)
library(ReactomePA)
library(gridExtra)
library(cowplot)

library(conflicted)

conflict_prefer("rename", "dplyr")
conflict_prefer("select", "dplyr")
```

##Sorting the metadata
```{r}
anno <- read_csv2(paste0(params$data, "/metadata_PASeq.csv"))

anno <- anno |>
  mutate(seq_ID = gsub("AK19_", "", seq_ID)) |>
  relocate(c("place", "seq_ID", "patient", "GW", "sectio", "GDM", "Grav", "Para", "weight", "Csec", "PA_score", "Placenta_previa")) |>
  mutate_at(vars(-c("seq_ID", "week", "day", "GW", "weight")), factor) |>
  mutate(sectio = recode_factor(sectio,
                                "Prim. Sectio"= "1sec",
                                "Prim. sectio"= "1sec",
                                "Sek. Sectio"= "2sec")) |>
  mutate(place = recode_factor(place, "control" = "Ctrl", "control_mus" = "CtrlMus")) |>
  mutate(place = fct_relevel(place, c("Ctrl", "PAS", "CtrlMus")))

# Add modified covariats to metadata
anno$ScaledGW <- scale(anno$GW, center = TRUE)
anno$ScaledWeight <- scale(anno$weight, center = TRUE)

# Transform continuous covariates into factors by cutting them (Ideally factor levels should make sense biologically!")
anno$cutGW <- cut(anno$GW, c(0, 243, 269, Inf), c("<243", "244-269", ">269"), include.lowest = TRUE)
anno$cutWeight <- cut(anno$weight, c(0, 2000, 3201, Inf), c("<2000", "2001-3201", ">3201"), include.lowest = TRUE)
```


## Reading in the count table 

```{r}
PAS <- read_tsv(paste0(params$data, "/PAS_counts.tsv"))

PAS <- PAS |>
  mutate(gene_id = gsub("\\.[0-9]*$", "", gene_id)) |> #regex pattern
  select(-gene_name) |>
  mutate_at(vars(-("gene_id")), ceiling) |>
  rename_with(~str_replace(., pattern = "AK19_", replacement = "")) |>
  column_to_rownames(var = "gene_id")
```
## Deseq and DE analysis

In the design we included Grav, Para, sectio, Csec and place

```{r}

dds <- DESeqDataSetFromMatrix(countData = PAS,
                              colData = anno |> column_to_rownames(var = "seq_ID"),
                              design = ~Grav + Para + sectio + Csec + place)
#prefiltering- removing rows with few reads
keep<-rowSums(counts(dds)) >=10
dds<-dds[keep,]
#saveRDS(dds, file="dds_mixeddesign.rds")
```

# running the Differential expression analysis

## PAS vs ctrl

note! for levels, the R will choose based on alphabetical order! careful 

```{r}

dds$place<-factor(dds$place, levels= c("Ctrl", "PAS", "CtrlMus"))
dds <- DESeq(dds)
results_PAS_ctrl <- results(dds)
results_PAS_ctrl <- results(dds, name ="place_PAS_vs_Ctrl")
results_PAS_ctrl <- results(dds, filterFun = ihw, contrast = c("place", "PAS", "Ctrl"))

results_PAS_ctrl<- as.data.frame(results_PAS_ctrl)
results_PAS_ctrl<-results_PAS_ctrl[order(results_PAS_ctrl$padj),]
sig.res<-results_PAS_ctrl|>  # using just the signif DEG
  subset(padj<0.05) 


```

## mus vs. ctrl

```{r}

results_mus_ctrl<-results(dds, filterFun= ihw, contrast = c("place", "CtrlMus", "Ctrl"))
results_mus_ctrl<- as.data.frame(results_mus_ctrl)
results_mus_ctrl<- results_mus_ctrl %>%
  subset(padj<0.05)
```

## pas vs mus

```{r}
results_mus_PAS<-results(dds, filterFun= ihw, contrast = c("place", "CtrlMus", "PAS"))
results_mus_PAS<- as.data.frame(results_mus_PAS)
results_mus_PAS<- results_mus_PAS %>%
  subset(padj<0.05)
```



```{r}
#merging the df into ENS_sig (between PAS and ctrl)
ENS_sig<- merge(as.data.frame(sig.res), PASbackup, by.y="gene_id", by.x= "row.names")

ENS_bg<-merge(as.data.frame(results_mus_ctrl), PASbackup, by.y="gene_id", by.x= "row.names")
```

### nCount data transformation

```{r}
# Extracting transformed values
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(rld), 3)
rld_df<-as.matrix(assay(rld))
#rld_df<-as.data.frame(rld_df)

```

```{r}
# rlog and vst transformation
dds <- estimateSizeFactors(dds)
rld <- rlog(dds)
vst <- vst(dds, blind = TRUE)
fpm <- fpm(dds)
#fpkm
```

```{r}
plot_layout <- function(df, x_string, y_string, title) {
  ggplot(df, aes_string(x = x_string, y = y_string)) +
    geom_point(size = 0.3) +
    ggtitle(title)
}

# Have a look at different transformations by plotting first against second sample
p <- plot_grid(plot_layout(as_tibble(log2(1 + counts(dds, normalized = TRUE))),"TG.1", "TG.2", "log2"),
               plot_layout(as_tibble(assay(rld)),"TG.1", "TG.2", "rlog"),
               plot_layout(as_tibble(assay(vst)),"TG.1", "TG.2", "vst"),
               plot_layout(as_tibble(log2(fpm)),"TG.1", "TG.2", "fpm"))
```
### likelihood test ratio

[Likelihood test ratio workfloq](https://hbctraining.github.io/DGE_workshop_salmon/lessons/08_DGE_LRT.html)

```{r}

# perform likelihood ratio test
dds_lrt <- DESeq(dds, test = "LRT",
                 full = ~Grav + Para + sectio + Csec + place,
                 reduced = ~Grav + Para + sectio + Csec)

res_lrt <- results(dds_lrt)
    
summary(res_lrt)
sum(res_lrt$padj < 0.01, na.rm = TRUE)
sum(res_lrt$padj < 0.05, na.rm = TRUE)

# Create a tibble for LRT results
res_LRT_tb <- res_LRT %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

# Subset to return genes with padj < 0.05
sigLRT_genes <- res_LRT_tb %>% 
  dplyr::filter(padj < 0.05)

# Get number of significant genes
nrow(sigLRT_genes)

```

# Clustering the genes

Aim: To identify gene clusters exhibiting particular patterns
Results: 
Cluster 2: genes that are upregulated in ctrl vs. PAS


```{r}
 #Subset results for faster cluster finding (for classroom demo purposes)
clustering_sig_genes <- sigLRT_genes %>%
  arrange(padj) %>%
  head(n=1000)


# Obtain rlog values for those significant genes
rdl_mat<-assay(rld)
cluster_rlog <- rdl_mat[clustering_sig_genes$gene, ]

#you have to put the anno in factors
anno$place<-as.factor(anno$place)

# determining sets of genes with similar expression patterns across sample groups
# Use the `degPatterns` function from the 'DEGreport' package to show gene clusters across sample groups
#BiocManager::install("DEGreport")
#library(DEGreport)
clusters <- degPatterns(cluster_rlog, 
                        metadata = anno, 
                        time = "place", 
                        col=NULL ) 
#??? put the gene names in the df
head(clusters$df)

cluster_2_genes <- clusters$df %>%
          filter(cluster == 2)
```

```{r}
plotPCA(vsd, intgroup = "place", ntop = 500) + 
  geom_text(aes(label = name),vjust = 2)
#color on 2 levels place and PAscore
```

## Plot pca using ggplot
```{r}

rld <- rlog(dds)
vst

# Extract pca coordinates
pcaData <- plotPCA(rld, intgroup = c("place", "PA_score"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"), digits = 1)

p <- ggplot(pcaData, aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = PA_score),
             size = 8,
             shape = 21) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggtitle("PCA Transcriptome")
```

```{r}
p <- ggplot(pcaData, aes(x=PC1, y=PC2, col=place)) + 
  geom_point(aes(shape=PA_score), size = 8) + 
  labs(shape = "PA_score", colour = "place", size = "PA_score")
```

<https://bioconductor.org/packages/release/bioc/manuals/ReactomePA/man/ReactomePA.pdf> \## ReactomePA
