---
title: "Exploration_data"
author: "Tina Gorsek Sparovec"

params:
data: "../4.0_data"
results: "../2.0_results"
lib: "../0.0_lib"
maxcores: 6
output:
html_notebook:
theme: spacelab
highlight: textmate
toc: yes
number_sections: true
toc_depth: 3
toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Workflow sources: 
[Analyzing RNA-seq data with DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat)

```{r}
#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("pasilla")
library(pasilla)
library(DESeq2)
library(tidyverse)

#BiocManager::install(c("IHW","pcaExplorer","clusterProfiler","ReactomePA"))
library(IHW)
library(pcaExplorer)
library(clusterProfiler)
library(ReactomePA)


```


#Meta data wrangling

```{r}
anno <- read.csv2(paste0(params$data,"/metadata_PASeq.csv"),
                  sep= ";")

anno <- anno |>
  mutate(seq_ID = gsub("AK19_", "", seq_ID)) |> #renaming the seq_ID
  relocate(c("place", "seq_ID", "patient", "GW", "sectio", "GDM", "Grav", "Para", "weight", "Csec", "PA_score", "Placenta_previa")) |>
  #change column order/ select columns
  mutate_at(vars(-c("seq_ID", "week", "day", "GW", "weight")), factor) |> 
  #assigning all except the mentioned in factors
  mutate(sectio = recode_factor(sectio,
                                "Prim. Sectio"= "1sec", 
                                "Prim. sectio"= "1sec", 
                                "Sek. Sectio"= "2sec")) |>
  #changing the variables into factors
  mutate(place = recode_factor(place,
                               "control"= "Ctrl", 
                               "control_mus"= "CtrlMus", 
                               "PAS"= "PAS" )) |>
  mutate(place = fct_relevel(place, c("Ctrl", "PAS", "CtrlMus")))
  
# If you have to mutate all columns- use mutate_all(factor)

head(anno)


# Add scaled covariated to metadata
anno$GWscaled<-scale(anno$GW, center=TRUE)
anno$SWeight<-scale(anno$weight, center = TRUE)

# Transforming the continuous variables into factors by cutting them (should make biological sense)
anno$cutGW <- cut(anno$GW, 
                  c(0, 196, 224, 259, Inf), 
                  c("<28","28-32", "32-37", ">38"), 
                  include.lowest = TRUE)
#cutting the GW on extremely preterm < 28 weeks, Preterm 28-32, Moderate to Late preterm 32-37, Normal 38-42


anno$cutWeight <- cut(anno$weight, 
                      c(0, 1500, 2500, 4200, Inf), 
                      c("VLBW", "LBW", "NBW", "LGW"), 
                      include.lowest = TRUE)
# cutting the Weight in very low birth weight (VLBW) <1500, low birth weight (LBW) <2500, Normal birth weight (NBW) 2500- 4200, Large for gastational weight > 4200g

#saveRDS(anno, file= "anno.rds")
```

[reference birth weight](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5710996/)
[reference gastational weeks](https://medlineplus.gov/ency/article/002367.htm)

## Reading in the count table 

```{r}
PAS<-read.tsv(paste0(params$data, "/PAS_counts.tsv"), sep="", header = T)


PASbackup<-PAS
PAS<-PAS[,-2]
names<-make.unique(PAS$gene_id) #making the column name gene_names as rown names
rownames(PAS)<-names
PAS<-PAS[,-1]

# rounding up the PAS counts
PAS<-ceiling(PAS[,])

#saveRDS(PAS, file="PAS.rds")
```

## Deseq and DE analysis

In the design we included Grav, Para, sectio, Csec and place

```{r}
dds<-DESeqDataSetFromMatrix(countData = PAS,
                            colData=anno,
                            design = ~ Grav + Para + sectio + Csec + place)
#prefiltering- removing rows with few reads
keep<-rowSums(counts(dds)) >=10
dds<-dds[keep,]
#saveRDS(dds, file="dds_mixeddesign.rds")
```

# running the Differential expression analysis

## PAS vs ctrl

note! for levels, the R will choose based on alphabetical order! careful 

```{r}

dds$place<-factor(dds$place, levels= c("ctrl", "mus", "PAS"))
dds <- DESeq(dds)
results_PAS_ctrl <- results(dds)
results_PAS_ctrl<-results(dds, name="place_PAS_vs_ctrl")
results_PAS_ctrl<-results(dds, filterFun= ihw, contrast = c("place", "PAS", "ctrl"))

results_PAS_ctrl<- as.data.frame(results_PAS_ctrl)
results_PAS_ctrl<-results_PAS_ctrl[order(results_PAS_ctrl$padj),]
sig.res<-results_PAS_ctrl|>  # using just the signif DEG
  subset(padj<0.05) 


```

## mus vs. ctrl

```{r}

results_mus_ctrl<-results(dds, filterFun= ihw, contrast = c("place", "mus", "ctrl")) # reference level comes last!
results_mus_ctrl<- as.data.frame(results_mus_ctrl)
results_mus_ctrl<- results_mus_ctrl %>%
  subset(padj<0.05)
```

## pas vs mus

```{r}
results_mus_PAS<-results(dds, filterFun= ihw, contrast = c("place", "PAS", "mus"))
results_mus_PAS<- as.data.frame(results_mus_PAS)
results_mus_PAS<- results_mus_PAS %>%
  subset(padj<0.05)
```



```{r}
#merging the df into ENS_sig (between PAS and ctrl)
ENS_sig<- merge(as.data.frame(sig.res), PASbackup, by.y="gene_id", by.x= "row.names") # have some kind of conversion key dataframe- at the very beggining just take the gene id and symbols into df, appent this - so that you dont have the counts.

ENS_bg<-merge(as.data.frame(results_mus_ctrl), PASbackup, by.y="gene_id", by.x= "row.names")
```

### nCount data transformation

```{r}
# Extracting transformed values
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(rld), 3)
rld_df<-as.matrix(assay(rld))
#rld_df<-as.data.frame(rld_df)
### plot the first against the second sample- if they are on the same line 
```

### likelihood test ratio

[Likelihood test ratio workfloq](https://hbctraining.github.io/DGE_workshop_salmon/lessons/08_DGE_LRT.html)

```{r}
dds_lrt <- DESeq(dds, test="LRT", reduced = ~ 1)
res_LRT <- results(dds_lrt)

# Create a tibble for LRT results
res_LRT_tb <- res_LRT %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

# Subset to return genes with padj < 0.05
sigLRT_genes <- res_LRT_tb %>% 
  dplyr::filter(padj < 0.05)

# Get number of significant genes
nrow(sigLRT_genes)

```

# Clustering the genes

Aim: To identify gene clusters exhibiting particular patterns
Results: 
Cluster 2: genes that are upregulated in ctrl vs. PAS


```{r}
 #Subset results for faster cluster finding (for classroom demo purposes)
clustering_sig_genes <- sigLRT_genes %>%
  arrange(padj) %>%
  head(n=1000)


# Obtain rlog values for those significant genes
rdl_mat<-assay(rld)
cluster_rlog <- rdl_mat[clustering_sig_genes$gene, ]

#you have to put the anno in factors
anno$place<-as.factor(anno$place)

# determining sets of genes with similar expression patterns across sample groups
# Use the `degPatterns` function from the 'DEGreport' package to show gene clusters across sample groups
#BiocManager::install("DEGreport")
library(DEGreport)
clusters <- degPatterns(cluster_rlog, 
                        metadata = anno, 
                        time = "place", 
                        col=NULL ) 
#??? put the gene names in the df
head(clusters$df)

cluster_2_genes <- clusters$df %>%
          filter(cluster == 2)
```

```{r}
plotPCA(rld, intgroup="place", ntop=500)+geom_text(aes(label=name),vjust=2)
#color on 2 levels place and PAscore
```



<https://bioconductor.org/packages/release/bioc/manuals/ReactomePA/man/ReactomePA.pdf> \## ReactomePA
