---
title: "Metadata_wrangling_DESeq"
author: "Tina Gorsek Sparovec"

params:
 data: "../data"
 results: "../res"
 lib: "../lib"
 rds: "../rds"
 maxcores: 6
output:
 html_notebook:
 theme: spacelab
 highlight: textmate
 toc: yes
 number_sections: true
 toc_depth: 3
 toc_float: true
---



Workflow sources: 
[Analyzing RNA-seq data with DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat)


```{r}
library(DESeq2)
library(tidyverse)
library(readr)
library(IHW)
library(pcaExplorer)
library(clusterProfiler)
library(ReactomePA)
library(DEGreport)
library(cowplot)
library(conflicted)

conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
```


#Meta data wrangling

```{r}
anno <- read.csv2(paste0(params$data,"/metadata_PASeq2.csv"),
                  sep= ";")

anno <- anno |>
  filter(seq_ID != "AK19_TG.5") |> 
  filter(seq_ID != "AK19_TG.8") |>
  filter(seq_ID != "AK19_TG.11") |> #removing an outlier 
  mutate(seq_ID = gsub("AK19_", "", seq_ID)) |> #renaming the seq_ID
  relocate(c("place", "seq_ID", "patient", "GW", "sectio", "GDM", "Grav", "Para", "weight", "Csec", "PA_score", "Placenta_previa")) |>
  #change column order/ select columns
  mutate_at(vars(-c("seq_ID", "week", "day", "GW", "weight")), factor) |> 
  #assigning all except the mentioned in factors
  mutate(sectio = recode_factor(sectio,
                                "Prim. Sectio"= "1sec", 
                                "Prim. sectio"= "1sec", 
                                "Sek. Sectio"= "2sec")) |>
  #changing the variables into factors
  mutate(place = recode_factor(place,
                               "control"= "Ctrl", 
                               "control_mus"= "CtrlMus", 
                               "PAS"= "PAS" )) |>
  mutate(place = fct_relevel(place, c("Ctrl", "PAS", "CtrlMus"))) 
  
  
# If you have to mutate all columns- use mutate_all(factor)

head(anno)


# Add scaled covariated to metadata
anno$GWscaled<-scale(anno$GW, center=TRUE)
anno$SWeight<-scale(anno$weight, center = TRUE)

# Transforming the continuous variables into factors by cutting them (should make biological sense)
anno$cutGW <- cut(anno$GW, 
                  c(0, 196, 224, 259, Inf), 
                  c("<28","28-32", "32-37", ">38"), 
                  include.lowest = TRUE)
###cutting the GW on extremely preterm < 28 weeks, Preterm 28-32, Moderate to Late preterm 32-37, Normal 38-42


anno$cutWeight <- cut(anno$weight, 
                      c(0, 1500, 2500, 4200, Inf), 
                      c("VLBW", "LBW", "NBW", "LGW"), 
                      include.lowest = TRUE)
###cutting the Weight in very low birth weight (VLBW) <1500, low birth weight (LBW) <2500, Normal birth weight (NBW) 2500- 4200, Large for gastational weight > 4200g

#saveRDS(anno, file = file.path(params$rds, "anno_noout.rds"))
```

[reference birth weight](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5710996/)
[reference gastational weeks](https://medlineplus.gov/ency/article/002367.htm)

## Reading in the count table 

```{r}
PAS<-read_tsv(paste0(params$data, "/PAS_counts.tsv"))

#Object with gene_id and gene_names
KEY <- PAS |>
  select(-c(AK19_TG.5, AK19_TG.8, AK19_TG.11)) |>
  select(gene_id, gene_name) |>
  mutate(gene_id = gsub("\\.[1-9]*$", "", gene_id))

#data wrangling
PAS <- PAS |> 
  select(-c(AK19_TG.5, AK19_TG.8, AK19_TG.11)) |>
  mutate(gene_id = gsub("\\.[1-9]*$", "", gene_id)) |> 
  select(-gene_name) |>
  mutate_at(vars(-("gene_id")), ceiling) |> #rounding the counts
  rename_with(~str_replace(., pattern = "AK19_", replacement = "")) |>
  column_to_rownames(var = "gene_id") 


#saveRDS(PAS, file = file.path(params$rds,"PAS_noout.rds"))
```

## Deseq and DE analysis

In the design we included Grav, Para, Csec and place
Variables: weight, 

<!-- VJM: You can usually play around here and add/remove explanatory variables. For the final analysis I would suggest removing the terms that don't explain much variance! -->

```{r}
dds<-DESeqDataSetFromMatrix(countData = PAS,
                            colData = anno |> column_to_rownames(var = "seq_ID"),
                            design = ~ Grav + Para + Csec + place)

#prefiltering- removing rows with few reads
keep<-rowSums(counts(dds)) >=10
dds<-dds[keep,]

#saveRDS(dds, file = file.path(params$rds, "dds_mixeddesign_noout.rds"))

data <- plotCounts(dds, gene= "ENSG00000000003",
           intgroup = "place",
           returnData = TRUE)
ggplot(data, aes(x = place, y= count))+
  geom_point(aes(fill=place))
  
```

# running the Differential expression analysis

## PAS vs ctrl

note! for levels, the R will choose based on alphabetical order! careful 
*contrast* specifies the comparison between the 2 levels of a variable

```{r}

dds$place<-factor(dds$place, 
                  levels= c("Ctrl", "CtrlMus", "PAS"))
dds <- DESeq(dds)


results_PAS_ctrl <-results(dds, 
                           filterFun = ihw, 
                           name="place_PAS_vs_Ctrl",
                           contrast = c("place", "PAS", "Ctrl"))
# what about the log for shrinkage? 
results_PAS_ctrl <- as.data.frame(results_PAS_ctrl)
results_PAS_ctrl <-results_PAS_ctrl[order(results_PAS_ctrl$padj),]
sig.res <- results_PAS_ctrl|>  
  subset(padj<0.05)


#<!-- VJM: padj is not equal to pval! Your comment "signif DEG" is a bit ambiguous -->

```
Results:
*baseMean* average of normalised count values divided by size factors
*log2FoldChange* effect size estimate - how much of the gene's expression is different between the conditions
*pajd* - we consider 10% of false positive acceptable (FDR)

## mus vs. ctrl

```{r}
results_mus_ctrl<-results(dds, 
                          filterFun= ihw, 
                          contrast = c("place", "CtrlMus", "Ctrl")) # reference level comes last!
results_mus_ctrl<- as.data.frame(results_mus_ctrl)
results_mus_ctrl <-results_mus_ctrl[order(results_mus_ctrl$padj),]
results_mus_ctrl<- results_mus_ctrl |> 
  subset(padj < 0.05) 
```

## pas vs mus

```{r}

results_mus_PAS<-results(dds, filterFun= ihw, contrast = c("place", "PAS", "CtrlMus"))
results_mus_PAS<- as.data.frame(results_mus_PAS)
results_mus_PAS<- results_mus_PAS |> 
  subset(padj < 0.05)
```



```{r}
#Merging the df into ENS_sig (between PAS and ctrl)
#[]residual must be the same as ens bg
ENS_sig <- merge(as.data.frame(sig.res), 
                 KEY, 
                 by.y="gene_id", 
                 by.x= "row.names") # have some kind of conversion key dataframe- at the very begining just take the gene id and symbols into df, appent this - so that you dont have the counts.

ENS_bg <- merge(as.data.frame(results_mus_ctrl), 
                KEY, 
                by.y="gene_id", 
                by.x= "row.names")
ENS_bg <- ENS_bg |> 
   subset(padj < 0.05) 

#excluding the rows from the ENS_sig, that appear in the ENS_bg and ENS_sig
ENS_sig_all <- ENS_sig |> 
   subset(padj < 0.05) |>
  pull(gene_name)

ENS_sig_names <- ENS_sig |>
  subset(!ENS_sig$Row.names %in% ENS_bg$Row.names) |> 
  pull(gene_name)


#<!-- VJM: Removing the ctrlMus genes from the PAS genes is just one idea to get rid of the background. If it doesn't yield many hits I would try to run the downstream analysis with the unfiltered PAS list. Afterwards you can still check what genes were also in the ctrlMus and remove the hits that might not be worth looking at more closely -->


```

### nCount data transformation

*regularized-logarithm transformation*, or rlog for short. For genes with high counts, the rlog transformation differs not much from an ordinary log2 transformation. For genes with lower counts, however, the values are shrunken towards the genesâ€™ averages across all samples. Using an empirical Bayesian prior in the form of a ridge penalty, this is done such that the rlog-transformed data are approximately homoskedastic.
[source](http://www.sthda.com/english/wiki/rna-seq-differential-expression-work-flow-using-deseq2)

```{r}

# Extracting transformed values
vst <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(rld), 3)
rld_df <- as.matrix(assay(rld))
fpm <- fpm(dds)

## Results of each transformation plotted 
plot_layout <- function(df, x, y, title) {
  ggplot(df, aes_string(x = x, y = y)) + 
    geom_point(size = 0.2) +
    ggtitle(title)
}
plot <- plot_grid(plot_layout(as_tibble(log2(1 + counts(dds, normalized =T))), "TG.10", "TG.12", "log2"),
                  plot_layout(as_tibble(assay(rld)), "TG.10", "TG.12", "rld"),
                  plot_layout(as_tibble(assay(vst)), "TG.10", "TG.12", "vst"),
                  plot_layout(as_tibble(log2(fpm)), "TG.10", "TG.12", "fpm"))
plot(plot)
```
Results: we want to have a transformation that gives us the least outliers. In this case rlog transformation if the best choice. Rlog compresses the differences for genes.


# Clustering the genes

Aim: To identify gene clusters exhibiting particular patterns
Results: 
Cluster 3: genes that are downregulated in PAS vs ctrl
Cluster 4: genes that are upregulated in PAS vs. ctrl

<!-- VJM: I would suggest using the DEGs from the "normal" comparison PAS vs. Ctrl for the downstream analysis!!! Taking the LRT comparison because there are more significant genes to work with is going to waste your time - and potentially other peoples time - chasing false positive results. You should start with a question/hypothesis and stick with it !! -->

```{r}

#degPatterns using rld results

rld2 <-rld[rownames(rld) %in% rownames(results_PAS_ctrl|> subset(padj < 0.05))]
rld3 <- as_tibble(assay(rld2[rownames(rld2) %in% rownames(results_PAS_ctrl|> subset(padj < 0.05))])) |>
  mutate(id = rownames(rld2)) |>
  column_to_rownames(var = "id")

clusters <- degPatterns(rld3, 
                        metadata = anno |> 
                          column_to_rownames(var = "seq_ID"),
                        time = "place", 
                        col = NULL)

#Extracting the genes from the cluster
clust_1 <- 1
clust_3 <- 3

cluster_2_genes <- clusters$df |>
  rownames_to_column(var = "gene_id") |>
  left_join(KEY) |>
  relocate(gene_name, .after = gene_id) |>
  filter(cluster == clust_3) |>
  select(-genes)
```


## Plot PCA 

Comment on the PCA = PA_score needs to be adjusted, CtrlMus uses the PA_Score from the patients with PAS, however it is just a control. 

```{r}

plotPCA(rld, intgroup="place", ntop=500) +
  geom_text(aes(label=name),vjust=2)
#color on 2 levels place and PAscore

#Plot the PCA using ggplot

#Extract PCA coordinates
pcaData <- plotPCA(rld, 
                   intgroup = c("place", "PA_score"), 
                   returnData = TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"), digits = 1)

p <- ggplot(pcaData, 
            aes(x = PC1, y = PC2, color = place)) +
  geom_point(aes(fill = place), size=3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggtitle("PCA") +
  scale_color_manual(values = c("#fc8961","#b73779", "#180f3d")) +
  theme_bw()
  
```
### Different visualisation
```{r}
p2 <- ggplot(pcaData,
            aes(x = PC1, y = PC2)) +
  geom_point(aes(shape = PA_score), size = 5) +
  labs(shape = "PA_score", colour = "place", size = "PA_score")
```
