---
title: "Metadata_wrangling_DESeq"
author: "Tina Gorsek Sparovec"

params:
 data: "../data"
 results: "../res"
 lib: "../lib"
 maxcores: 6
output:
 html_notebook:
 theme: spacelab
 highlight: textmate
 toc: yes
 number_sections: true
 toc_depth: 3
 toc_float: true
---



Workflow sources: 
[Analyzing RNA-seq data with DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat)


```{r}
library(pasilla)
library(DESeq2)
library(tidyverse)
library(readr)
library(IHW)
library(pcaExplorer)
library(clusterProfiler)
library(ReactomePA)
library(DEGreport)
library(cowplot)
```


#Meta data wrangling

```{r}
anno <- read.csv2(paste0(params$data,"/metadata_PASeq.csv"),
                  sep= ";")

anno <- anno |>
  filter(seq_ID != "AK19_TG.8") |> #removing an outlier 
  mutate(seq_ID = gsub("AK19_", "", seq_ID)) |> #renaming the seq_ID
  relocate(c("place", "seq_ID", "patient", "GW", "sectio", "GDM", "Grav", "Para", "weight", "Csec", "PA_score", "Placenta_previa")) |>
  #change column order/ select columns
  mutate_at(vars(-c("seq_ID", "week", "day", "GW", "weight")), factor) |> 
  #assigning all except the mentioned in factors
  mutate(sectio = recode_factor(sectio,
                                "Prim. Sectio"= "1sec", 
                                "Prim. sectio"= "1sec", 
                                "Sek. Sectio"= "2sec")) |>
  #changing the variables into factors
  mutate(place = recode_factor(place,
                               "control"= "Ctrl", 
                               "control_mus"= "CtrlMus", 
                               "PAS"= "PAS" )) |>
  mutate(place = fct_relevel(place, c("Ctrl", "PAS", "CtrlMus"))) 
  
  
# If you have to mutate all columns- use mutate_all(factor)

head(anno)


# Add scaled covariated to metadata
anno$GWscaled<-scale(anno$GW, center=TRUE)
anno$SWeight<-scale(anno$weight, center = TRUE)

# Transforming the continuous variables into factors by cutting them (should make biological sense)
anno$cutGW <- cut(anno$GW, 
                  c(0, 196, 224, 259, Inf), 
                  c("<28","28-32", "32-37", ">38"), 
                  include.lowest = TRUE)
###cutting the GW on extremely preterm < 28 weeks, Preterm 28-32, Moderate to Late preterm 32-37, Normal 38-42


anno$cutWeight <- cut(anno$weight, 
                      c(0, 1500, 2500, 4200, Inf), 
                      c("VLBW", "LBW", "NBW", "LGW"), 
                      include.lowest = TRUE)
###cutting the Weight in very low birth weight (VLBW) <1500, low birth weight (LBW) <2500, Normal birth weight (NBW) 2500- 4200, Large for gastational weight > 4200g

#saveRDS(anno, file = "anno.rds")
```

[reference birth weight](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5710996/)
[reference gastational weeks](https://medlineplus.gov/ency/article/002367.htm)

## Reading in the count table 

```{r}
PAS<-read_tsv(paste0(params$data, "/PAS_counts.tsv"))

#Object with gene_id and gene_names
KEY <- PAS |>
  select(-AK19_TG.8) |>
  select(gene_id, gene_name) |>
  mutate(gene_id = gsub("\\.[1-9]*$", "", gene_id))

#data wrangling
PAS <- PAS |> 
  select(-AK19_TG.8) |>
  mutate(gene_id = gsub("\\.[1-9]*$", "", gene_id)) |> 
  select(-gene_name) |>
  mutate_at(vars(-("gene_id")), ceiling) |> #rounding the counts
  rename_with(~str_replace(., pattern = "AK19_", replacement = "")) |>
  column_to_rownames(var = "gene_id") 


#saveRDS(PAS, file="PAS.rds")
```

## Deseq and DE analysis

In the design we included Grav, Para, sectio, Csec and place

```{r}
dds<-DESeqDataSetFromMatrix(countData = PAS,
                            colData = anno |> column_to_rownames(var = "seq_ID"),
                            design = ~ Grav + Para + sectio + Csec + place)

#prefiltering- removing rows with few reads
keep<-rowSums(counts(dds)) >=10
dds<-dds[keep,]

#saveRDS(dds, file="dds_mixeddesign.rds")
```

# running the Differential expression analysis

## PAS vs ctrl

note! for levels, the R will choose based on alphabetical order! careful 

```{r}

dds$place<-factor(dds$place, 
                  levels= c("Ctrl", "CtrlMus", "PAS"))
dds <- DESeq(dds)
results_PAS_ctrl <- results(dds)
results_PAS_ctrl <-results(dds, 
                           name="place_PAS_vs_Ctrl")
results_PAS_ctrl <-results(dds, 
                           filterFun = ihw, 
                           contrast = c("place", "PAS", "Ctrl"))

results_PAS_ctrl <- as.data.frame(results_PAS_ctrl)
results_PAS_ctrl <-results_PAS_ctrl[order(results_PAS_ctrl$padj),]
sig.res <- results_PAS_ctrl|>  # using just the signif DEG
  subset(padj<0.05) 


```

## mus vs. ctrl

```{r}

results_mus_ctrl<-results(dds, 
                          filterFun= ihw, 
                          contrast = c("place", "CtrlMus", "Ctrl")) # reference level comes last!
results_mus_ctrl<- as.data.frame(results_mus_ctrl)
results_mus_ctrl <-results_mus_ctrl[order(results_mus_ctrl$padj),]
results_mus_ctrl<- results_mus_ctrl %>%
  subset(padj<0.05)
```

## pas vs mus

```{r}

results_mus_PAS<-results(dds, filterFun= ihw, contrast = c("place", "PAS", "CtrlMus"))
results_mus_PAS<- as.data.frame(results_mus_PAS)
results_mus_PAS<- results_mus_PAS %>%
  subset(padj<0.05)
```



```{r}
#Merging the df into ENS_sig (between PAS and ctrl)
ENS_sig <- merge(as.data.frame(sig.res), 
                 KEY, 
                 by.y="gene_id", 
                 by.x= "row.names") # have some kind of conversion key dataframe- at the very begining just take the gene id and symbols into df, appent this - so that you dont have the counts.

ENS_bg <- merge(as.data.frame(results_mus_ctrl), 
                KEY, 
                by.y="gene_id", 
                by.x= "row.names")
residual <- ENS_bg |>
  pull(Row.names)



```

### nCount data transformation

```{r}

# Extracting transformed values
vst <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
head(assay(rld), 3)
rld_df <- as.matrix(assay(rld))
fpm <- fpm(dds)

## Results of each transformation plotted 
plot_layout <- function(df, x, y, title) {
  ggplot(df, aes_string(x = x, y = y)) + 
    geom_point(size = 0.2) +
    ggtitle(title)
}
plot <- plot_grid(plot_layout(as_tibble(log2(1 + counts(dds, normalized =T))), "TG.10", "TG.11", "log2"),
                  plot_layout(as_tibble(assay(rld)), "TG.10", "TG.11", "rld"),
                  plot_layout(as_tibble(assay(vst)), "TG.10", "TG.11", "vst"),
                  plot_layout(as_tibble(log2(fpm)), "TG.10", "TG.11", "fpm"))
plot(plot)
```

## likelihood test ratio

[Likelihood test ratio workflow](https://hbctraining.github.io/DGE_workshop_salmon/lessons/08_DGE_LRT.html)

```{r likelihood test ratio}

dds_lrt <- DESeq(dds, 
                 test="LRT", 
                 full = ~ Grav + Para + sectio + Csec + place,
                 reduced = ~ Grav + Para + sectio + Csec )
res_LRT <- results(dds_lrt)

# Summary of results
summary(res_LRT)
sum(res_LRT$padj < 0.01, na.rm =TRUE)
sum(res_LRT$padj < 0.05, na.rm =TRUE)

# Create a tibble for LRT results
res_LRT_tb <- res_LRT |>
  data.frame() |>
  rownames_to_column(var="gene") |>
  as_tibble() |>
  filter(padj<0.05) #subsetting for the significant genes


# Get number of significant genes and compared it to the wald test 
nrow(res_LRT_tb)
nrow(ENS_sig)
```

# Clustering the genes

Aim: To identify gene clusters exhibiting particular patterns
Results: 
Cluster 2: genes that are upregulated in ctrl vs. PAS


```{r}
 #Subset results for faster cluster finding 
clustering_sig_genes <- res_LRT_tb |>
  arrange(padj) |>
  head(n=1000)


# Obtain rlog values for those significant genes
rdl_mat <- assay(rld)
cluster_rlog <- rdl_mat[clustering_sig_genes$gene, ]

#Rownames of anno should cluster_rlog column namees

anno2 <- anno |> 
  column_to_rownames("seq_ID")

# determining sets of genes with similar expression patterns across sample groups
# Use the `degPatterns` function from the 'DEGreport' package to show gene clusters across sample groups
#BiocManager::install("DEGreport")

clusters <- degPatterns(cluster_rlog, 
                        metadata = anno2, 
                        time = "place", 
                        col = NULL ) 




#degPatterns using rld results

rld2 <-rld[rownames(rld) %in% rownames(results_PAS_ctrl|> subset(padj < 0.05))]
rld3 <- as_tibble(assay(rld2[rownames(rld2) %in% rownames(results_PAS_ctrl|> subset(padj < 0.05))])) |>
  mutate(id = rownames(rld2)) |>
  column_to_rownames(var = "id")

clusters <- degPatterns(rld3, 
                        metadata = anno |> 
                          column_to_rownames(var = "seq_ID"), 
                        time = "place", 
                        col = NULL)

clusters$df
clusters$plot

cluster_4_genes <- clusters$df %>%
          filter(cluster == 4)

```




## Plot PCA 

Comment on the PCA = PA_score needs to be adjusted, CtrlMus uses the PA_Score from the patients with PAS, however it is just a control. 

```{r}

plotPCA(rld, intgroup="place", ntop=500)+geom_text(aes(label=name),vjust=2)
#color on 2 levels place and PAscore

#Plot the PCA using ggplot

#Extract PCA coordinates
pcaData <- plotPCA(rld, 
                   intgroup = c("place", "PA_score"), 
                   returnData = TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"), digits = 1)

p <- ggplot(pcaData, 
            aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = PA_score),
              size = 5,
              shape = 21) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggtitle("PCA")
  
```


### Different visualisation
```{r}
p <- ggplot(pcaData,
            aes(x = PC1, y = PC2)) +
  geom_point(aes(shape = PA_score), size = 5) +
  labs(shape = "PA_score", colour = "place", size = "PA_score")
```
