---
title: "Gene_anno"
author: "Tina Gorsek Sparovec"

date: 26.1.22
params:
 data: "../data"
 results: "../res"
 lib: "../lib"
 vign: FALSE
 maxcores: 6
output:
 html_notebook:
 theme: spacelab
 highlight: textmate
 toc: yes
 number_sections: true
 toc_depth: 3
 toc_float: true
---

```{r}
library(tidyverse)
require(org.Hs.eg.db)
library(janitor)
library(topGO)
library(DOSE)
library(pcaExplorer)
library(enrichplot)
library(hgu95av2.db)


# Required data: from notebook 1.0_metadata_exploration

# (1) subtraction/subset of the results_PAS_Ctrl and results_CtrlMus_Ctrl 

results_PAS_ctrl
 
# (2) Background genes: genes of results_CtrlMus_Ctrl

results_mus_ctrl
```
 
 
 
 ## TopGO vignette
 
 http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.515.8324&rep=rep1&type=pdf
 

# TopGO on the RNAseq results


The input should be a named vector of DEG genes. Used is ENS_sig object created in
1.0_metadata_exploration. (changed on 11.7.22)


```{r puting the genes as named vector}

residual <- results_mus_ctrl |>
  subset(padj < 0.05) |>
  rownames_to_column(var = "gene_id") |>
  pull(gene_id)

PAS_ctrl <- results_PAS_ctrl |>
  rownames_to_column(var = "gene_id")|>
  subset(!gene_id %in% residual)

genes <- PAS_ctrl |> 
  subset(padj < 0.05) |> 
  pull(gene_id)


up_genes <- PAS_ctrl |>
  subset(padj < 0.05 & PAS_ctrl$log2FoldChange > 0) |>
  pull(gene_id)

dn_genes <- PAS_ctrl |>
  subset(padj < 0.05 & PAS_ctrl$log2FoldChange < 0) |>
  pull(gene_id)

#Extracting the all genes

all_genes <- PAS_ctrl |>
  pull(gene_id)


#all genes expressed as a background
all_genes_names <- names(keep)

```

##topGotable on Biological pathways ("BF") and Molecular function ("MF")

Extract functional terms enriched in the DE genes, based on topGO

```{r}
 
Res <- lapply(c("BP", "MF"), function(ontology) {
  topgode <- topGOtable(DEgenes = genes,
                        BGgenes = all_genes,
                        ontology = ontology,
                        mapping = "org.Hs.eg.db",
                        geneID= "ENSEMBL"
                        )
  })


Res
#write.csv(data.frame(Res[1]), file.path(params$results, "Res_BP.csv"))
#write.csv(data.frame(Res[2]), file.path(params$results, "Res_MF.csv"))
```

 
## Adding a column with ENTREZ ID 
 
 
```{r ENTREZ}
PAS_ctrl$ENTREZID <- mapIds(org.Hs.eg.db,
                       keys=PAS_ctrl$gene_id, 
                        column="ENTREZID",
                        keytype="ENSEMBL",
                        multiVals="first")

universe <- unique(PAS_ctrl$ENTREZID)
genes_entrez <- PAS_ctrl |>
  subset(padj < 0.05) |> 
  pull(ENTREZID) |> 
  unique()
```
## KEEGG pathway over-representation analasis
 
[Biomedical knowledge mining using GOSemSim and clusterProfiler](https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-kegg.html)

KEGG is a collection of manually drawn pathway maps representing molecular interaction and reaction networks. These pathways cover a wide range of biochemical processes that can be divided into 7 broad categories.

Over Representation Analysis (ORA) (Boyle et al. 2004) is a widely used approach to determine whether known biological functions or processes are over-represented (= enriched) in an experimentally-derived gene list, e.g. a list of differentially expressed genes (DEGs).

```{r}

Kegg <- enrichKEGG(gene = genes_entrez,
                   organism = 'hsa',
                   universe = universe,
                   pvalueCutoff = 0.05)
head(Kegg)
Kegg <- setReadable(Kegg, org.Hs.eg.db, keyType = "ENTREZID")

#write.csv(Kegg, file.path(params$results, "Kegg_PAS_Ctrl_all.csv"))
```

## Reactome enrich analysis

Reactome uses hypergeometric model (one sided version of Fishers exact test) to asses whether the number of selected genes associated with a reactome pathway is larger then expected. 

--> you need to run this if you want to analyse up and downregulated genes for enrichPathway
up_genes_entrez <- PAS_ctrl |>
  subset(padj < 0.05 & PAS_ctrl$log2FoldChange > 0) |>
  pull(ENTREZID) |> 
  unique()

dn_genes_entrez <- PAS_ctrl |>
  subset(padj < 0.05 & PAS_ctrl$log2FoldChange < 0) |>
  pull(ENTREZID) |> 
  unique()
  
RESULTS: dn_genes- no significant pathways

```{r reactome}
R_PA<-enrichPathway(gene = genes_entrez, 
                    organism = "human",
                    universe = universe,
                    pvalueCutoff = 0.05, #had to change the pvaluecutoff  to 0.2because there was nothing to plot
                    readable = T)

PA_res <- as_tibble(R_PA@result)  


#write.csv(PA_res, file.path(params$results, "Reactome_PAS_Ctrl_all.csv"))
```
## Wiki pathway

WikiPathways is a continuously updated pathway database curated by a community of researchers and pathway enthusiasts. WikiPathways produces monthly releases of gmt files for supported organisms at data.wikipathways.org

```{r}
Wiki_pathway <- enrichWP(gene = genes_entrez, 
                         universe = universe,
                         organism = "Homo sapiens",
                         pvalueCutoff = 0.05)


#adding the ENTREZID as geneID
Wiki_pathway <- setReadable(Wiki_pathway, org.Hs.eg.db, keyType = "ENTREZID")
Wiki_pathway <- as_tibble(Wiki_pathway@result)

```
 [x] Try with just up an down regulated
 dn regulated- only one pathway
 up regulated - no sig. 
 [x]as background use all of the genes that were expressed (from dds class: DESeqDataSet --> Result: same as using the background genes 
dim: 22839 18 )-- > potentially could be significant.

## Finer control of GO term tests

Algorithms:
-ELIM - The elim method was design to be more conservative then the
classic method and therefore one expects the p-values returned by the former method 
are lower bounded by the p-values returned by the later method.

```{r}
ontology <- "BP" #(Biological Pathway)
#ontology <- "CC" #(Cellular Component)
#ontology <- "MF" #(Molecular Function)

# Decide test algorithm
algorithm <- "weight01"
#algorithm <- "classic"
#algorithm <- "elim"
#algorithm <- "weight"
#algorithm <- "lea"
#algorithm <- "parentchild"

# Define the statistical test used
# For tests based on gene counts
statistic <- "fisher"      # Fischer's exact test

#For tests based on gene scores or gene ranks
#statistic <- "ks"         # Kolmogorov-Smirnov test
#statistic <- "t"          # t-test

# For tests based on gene expression
#statistic <- "globaltest" # globaltest

# A list of up-regulated genes
# dn_genes
# up_genes
upList <- factor(as.integer(all_genes %in% genes))
names(upList) <- all_genes

upGOdata <- new("topGOdata", 
                ontology = ontology, 
                allGenes = upList, 
                geneSel = function(x)(x == 1), 
                nodeSize = 10, 
                annot = annFUN.org, 
                mapping = "org.Hs.eg.db", 
                ID = "ENSEMBL")


upRes <- runTest(upGOdata, 
                 algorithm = algorithm, 
                 statistic = statistic)

# Build up result table
up_tab <- GenTable(upGOdata, Pval = upRes, topNodes = 20)

# Update table with full GO term name
# Update up table 
up_tab$Term <- sapply(up_tab$"GO.ID", 
                      function(go) Term(GO.db::GOTERM[[go]]))
up_sigGenes <- sigGenes(upGOdata)

# Retrieve gene symbols for each GO from the test result
up_AnnoList <- lapply(up_tab$"GO.ID", 
                      function(x) as.character(unlist(genesInTerm(object = upGOdata, whichGO = x))))
up_SigList <- lapply(up_AnnoList, function(x) intersect(x, up_sigGenes))

#Coerce gene list to a comma-separated vector
up_tab$Genes <- sapply(up_SigList, paste, collapse = ",")
# Cbind first 5 up 
cbind(head(up_tab$Genes, 5))
```

## Visualisation

```{r}
#Bar plot PA_R
PA_selected_terms <- c("Keratinization", "Diseases of glycosylation", "Signaling by TGFB family members", "ECM proteoglycans")

PA_res_selected <- PA_res
PA_res_selected <- PA_res[PA_res_selected$Description %in% PA_selected_terms,]

ggplot(PA_res_selected, aes(x= Count, y = reorder(Description, -p.adjust), fill = p.adjust)) +
  geom_bar(stat = "identity", aes(-log2(qvalue))) +
  xlab("Gene count") +
  ylab("Description") +
  scale_fill_continuous(low = "#8c2981", high = "#fe9f6d") +
  theme_bw()
  


mutate(R_PA, qscore = -log(p.adjust, base = 10)) |>
    barplot(x="qscore")

#Dot plot 
dotplot(R_PA, showCategory=30) + 
  ggtitle("dotplot for ORA") +
  scale_fill_continuous(low = "#8c2981", high = "#fe9f6d")

#dotplot GO
up_tab$Pval <- as.numeric(up_tab$Pval)
up_tab <- up_tab[up_tab$Pval < 0.05,]
up_tab <- up_tab[,c("GO.ID","Term","Pval")]
up_tab

ntop <- 20
ggdata <- up_tab[1:ntop,]
ggdata$Term <- factor(ggdata$Term, levels = rev(ggdata$Term)) # fixes order
ggplot(ggdata,
  aes(x = Term, y = -log10(Pval), size = -log10(Pval), fill = -log10(Pval))) +

  expand_limits(y = 1) +
  geom_point(shape = 21) +
  scale_size(range = c(2.5,12.5)) +
  scale_fill_continuous(low = "#8c2981", high = "#fe9f6d") +

  xlab('') + ylab('Enrichment score') +
  labs(
    title = 'GO MF analysis') +
    #subtitle = 'Top 50 terms ordered by Kolmogorov-Smirnov p-value',
    #subtitle = 'Top 20 terms ordered by Fisher Exact p-value',
    #caption = 'Cut-off lines drawn at equivalents of p=0.05, p=0.01, p=0.001') +

  geom_hline(yintercept = c(-log10(0.05), -log10(0.01), -log10(0.001)),
    linetype = c("dotted", "longdash", "solid"),
    colour = c("black", "black", "black"),
    size = c(0.5, 1.5, 3)) +

  theme_bw(base_size = 24) +
  theme(
    legend.position = 'right',
    legend.background = element_rect(),
    plot.title = element_text(angle = 0, size = 16, face = 'bold', vjust = 1),
    plot.subtitle = element_text(angle = 0, size = 14, face = 'bold', vjust = 1),
    plot.caption = element_text(angle = 0, size = 12, face = 'bold', vjust = 1),

    axis.text.x = element_text(angle = 0, size = 12, face = 'bold', hjust = 1.10),
    axis.text.y = element_text(angle = 0, size = 12, face = 'bold', vjust = 0.5),
    axis.title = element_text(size = 12, face = 'bold'),
    axis.title.x = element_text(size = 12, face = 'bold'),
    axis.title.y = element_text(size = 12, face = 'bold'),
    axis.line = element_line(colour = 'black'),

    #Legend
    legend.key = element_blank(), # removes the border
    legend.key.size = unit(1, "cm"), # Sets overall area/size of the legend
    legend.text = element_text(size = 14, face = "bold"), # Text size
    title = element_text(size = 14, face = "bold")) +

  coord_flip()

```

To do:

variance partition package https://rdrr.io/bioc/variancePartition/f/inst/doc/variancePartition.pdf 

doing heatmap using the genes 
[x]cluster genes
[] pathway genes

[] finish updating color palette in viz
[] select the GO for cell prolif, cell-cell or cell-matrix communication, wound healing 





 